discovery/src/main/java/org/wildfly/extension/discovery/StaticProviderAddHandler.java

```diff
@@ -0,0 +1,116 @@
+
+import org.jboss.as.controller.AbstractAddStepHandler;
+import org.jboss.as.controller.OperationContext;
+import org.jboss.as.controller.OperationFailedException;
+import org.jboss.as.controller.capability.RuntimeCapability;
+import org.jboss.as.controller.operations.global.WriteAttributeHandler;
+import org.jboss.as.controller.registry.Resource;
+import org.jboss.dmr.ModelNode;
+import org.wildfly.discovery.AttributeValue;
+import org.wildfly.discovery.ServiceURL;
+import org.wildfly.discovery.impl.MutableDiscoveryProvider;
+import org.wildfly.discovery.impl.StaticDiscoveryProvider;
+import org.wildfly.discovery.spi.DiscoveryProvider;
+
+/**
+ * @author <a href="mailto:david.lloyd@redhat.com">David M. Lloyd</a>
+ */
+class StaticProviderAddHandler extends AbstractAddStepHandler {
+    private static final StaticProviderAddHandler INSTANCE = new StaticProviderAddHandler();
+
+    static StaticProviderAddHandler getInstance() {
+        return INSTANCE;
+    }
+
+    private StaticProviderAddHandler() {
+        super(new Parameters().addAttribute(StaticProviderDefinition.SERVICES));
+    }
+
+    public void execute(final OperationContext context, final ModelNode operation) throws OperationFailedException {
+        context.registerCapability(
+            RuntimeCapability.Builder
+                .of(DiscoveryExtension.DISCOVERY_PROVIDER_CAPABILITY, true, new MutableDiscoveryProvider())
+                .build().fromBaseCapability(context.getCurrentAddressValue()));
+        super.execute(context, operation);
+    }
+
+    protected void recordCapabilitiesAndRequirements(final OperationContext context, final ModelNode operation, final Resource resource) throws OperationFailedException {
+        // no operation
+    }
+
+    protected void performRuntime(final OperationContext context, final ModelNode operation, final Resource resource) throws OperationFailedException {
+        modifyRegistrationModel(context, operation);
+    }
+
+    static void modifyRegistrationModel(OperationContext context, ModelNode op) throws OperationFailedException {
+        WriteAttributeHandler.INSTANCE.execute(context, op);
+        context.addStep(op, StaticProviderAddHandler::modifyRegistration, OperationContext.Stage.RUNTIME);
+    }
+
+    static void modifyRegistration(OperationContext context, ModelNode op) throws OperationFailedException {
+        final MutableDiscoveryProvider mutableDiscoveryProvider = context.getCapabilityRuntimeAPI(DiscoveryExtension.DISCOVERY_PROVIDER_CAPABILITY, context.getCurrentAddressValue(), MutableDiscoveryProvider.class);
+        if (op.hasDefined(DiscoveryExtension.SERVICES)) {
+            final List<ModelNode> list = StaticProviderDefinition.SERVICES.resolveModelAttribute(context, op).asList();
+            final List<ServiceURL> services = new ArrayList<>();
+            for (ModelNode servicesNode : list) {
+                final ModelNode abstractTypeNode = servicesNode.get(DiscoveryExtension.ABSTRACT_TYPE);
```

Comment:
Best practice for all of these is to use AttributeDefinition.resolveModelAttribute(). It may or may not add any value versus a simple modelNode.get(...) but it's more resilient if things change in the future. (The resolveModelAttribute ensures expressions are resolved and validated if supported and that any default value for the attribute is returned if the attribute is undefined.)